ui/apps/platform/src/Containers/Collections/RuleSelector/AutoCompleteSelect.tsx

```diff
@@ -10,9 +14,33 @@ export type AutoCompleteSelectProps = {
     onChange: (value: string) => void;
     validated: ValidatedOptions;
     isDisabled: boolean;
+    autocompleteProvider?: (search: string) => CancellableRequest<string[]>;
+    entityType: SelectorEntityType;
```

Comment:
If the only reason for `entityType` prop is for `ResourceIcon` element, then instead of context versus props, the question is cost in code complexity versus benefit in user experience. Besides hierarchical organization deployments, namespaces, clusters, and form element labels, it seems likely that naming conventions for the 3 entity types are quite different.

Reply:
On the positive side of the coin, bravo for composition via `autocompleteProvider` prop.

Reply:
> If the only reason for entityType prop is for ResourceIcon element, then instead of context versus props, the question is cost in code complexity versus benefit in user experience. Besides hierarchical organization deployments, namespaces, clusters, and form element labels, it seems likely that naming conventions for the 3 entity types are quite different.

I'm not quite sure I understand.

Reply:
* The component did not need `entityType` prop to fetch the data, because encapsulated in `autocompleteProvider` function, correct?
* Unless my eyes missed something in the changed lines, the reason to add `entityType` prop is to render different resource icon?

Maybe I misunderstood the comment the the description about prop drilling;

> now all need access to the `entityType`

If this is an example which motivated the comment, then my counter question is instead of how to deal with increasing coupling between components, whether this change is pushing beyond a limit of how much coupling is too costly.

It looks like the the component was potentially reusable to become independent of container.

Reply:
Ah ok got it, yeah good point. Early on I was thinking the `RuleSelector` component would have been a good candidate to extract to `Components/PatternFly` if it had uses outside of this container. I think a lot of the tight coupling exists across everything under the `./RuleSelector` directory since they were originally all one component that I split apart for readability.

I think we can make this reusable again if we need to by pulling the autocomplete function up, and maybe making the error pieces less Formik specific.

Reply:
What caught my eye is asymmetry between `entityType` encapsulated for data but not for rendering.

If such specific rendering is impossible to avoid, then a more symmetrical solution is prop for option rendering callback function.